[user] 회원 검색 구현

LimdaeIl opened the pull request:
## #️⃣연관된 이슈
- 이슈 번호: #75 

## 변경 타입
- [ ] 신규 기능 추가/수정
- [ ] 리팩토링

## 변경 내용

- **to-be**(변경 후 설명을 여기에 작성)
  - [ ] TODO 쿼리디에스엘로 회원 동적 검색 구현

## 체크리스트
- [ ] 코드가 제대로 동작하는지 확인했습니다.
- [ ] 관련 테스트를 추가했습니다.
- [ ] 문서(코드, 주석, README 등)를 업데이트했습니다.

## 코멘트
- (추가적인 설명이나 코멘트가 필요한 경우 여기에 작성)

Commit:
[#75][user] feat: 쿼리디에스엘로 회원 검색 구현에 따른 디펜던시 추가하기

Commit:
[#75][user] feat: 페이징 정보 형식 설정하기

```
        "page": {
            "size": 10,
            "number": 0,
            "totalElements": 2,
            "totalPages": 1
        }
```

이렇게 나오도록 설정했다.

Commit:
[#75][auth] 리포지토리명 변경에 따라 수정하기

Commit:
[#75][user] feat: QueryDSL 환경 설정 파일 구성하기

Commit:
[#75][user] feat: 조회 성공 응답 추가하기

Commit:
[#75][user] feat: 조회 서비스 구현하기

공통 모듈에 구현된 것으로 사용하려고 했는데.. 실패해서 지금 방식대로 사용하자

Commit:
[#75][user] feat: QueryDSL 리포 생성 및 JPA 리포지토리 수정하기

2~3시간 넘게 해결하지 못한 문제가 있는 구간이다.

인터페이스와 구현체Impl 은 반드시 동일하게 하자...

회원 조회를 고객이 하면 안되므로 internal 안으로 넣었다.

HanaHww2 commented on user/src/main/java/com/faster/user/UserApplication.java:
```diff
@@ -2,7 +2,11 @@
+import org.springframework.data.web.config.EnableSpringDataWebSupport;
+import org.springframework.data.web.config.EnableSpringDataWebSupport.PageSerializationMode;
 
+@EnableSpringDataWebSupport(pageSerializationMode = PageSerializationMode.VIA_DTO)
```
이건 어떤 역할을 수행하는 코드인지 궁금합니다.

LimdaeIl replied:
```
PageModule$PlainPageSerializationWarning : Serializing PageImpl instances as-is is not supported, meaning that there is no guarantee about the stability of the resulting JSON structure!
	For a stable JSON structure, please use Spring Data's PagedModel (globally via @EnableSpringDataWebSupport(pageSerializationMode = VIA_DTO))
	or Spring HATEOAS and Spring Data's PagedResourcesAssembler as documented in https://docs.spring.io/spring-data/commons/reference/repositories/core-extensions.html#core.web.pageables.
```

이전에 직렬화 신경 안써도 잘 처리 해줬습니다..! 
근데 스프링 3.3 이상 버전 이상부터는 Page 대신 PagedModel 객체 사용하는 것을 **권장** 하는데, 
Spring Boot 개발팀에서 기본 Page의 사용이 더 좋다는 내부 의견에 따라 Page 직렬화도 지원하도록 제공하는 애너테이션입니다..! 

궁극적으로는 궁극적으로는 사용자가 Spring HATEOAS를 사용하여 Page인스턴스를 렌더링하는 플로우로 개발자들이 개발하기를 원한대요....! 아래 스프링 부트 #39797 링크 드립니다!

- https://github.com/spring-projects/spring-boot/pull/39797

HanaHww2 replied:
@LimdaeIl 설명 감사합니다. 그런데, Page 응답 때문인 거면, 저희 Page 공통 응답 객체로 한 번 파싱해서 사용해줘도 될 것 같다는 생각이 들어요. 공통 page 응답으로 처리해주실 수 있을까요?

LimdaeIl replied:
맞습니다! 그거 사용하려고 했는데 제가 엉뚱한 곳에서 몇 시간 동안 에러 붙잡고 있어서 적용을 못했네요.. ㅠㅠ 공통 Page 응답으로 다시 수정하겠습니다!

에러는 UserJpaRepository 이름과 UserJpaRepositoryImpl 앞의 이름이 동일해야 하는데 에러 메시지도 예상도 안되게 파라미터나 sql 알 수 없다는 식으로만 나와서 오후 다 날렸네요..

HanaHww2 replied:
무슨 에러였는지 감도 안오네요... 수정하시느라 고생하셨습니다.🥲

HanaHww2 commented on user/src/main/java/com/faster/user/app/user/presentation/internal/InternalUserController.java:
```diff
@@ -0,0 +1,69 @@
+      Pageable pageable,
+      @RequestParam(value = "username", defaultValue = "", required = false) String username,
+      @RequestParam(value = "name", defaultValue = "", required = false) String name,
+      @RequestParam(value = "slackId", defaultValue = "", required = false) String slackId,
```
컨벤션에 따라 slack-id 로 name 수정해주시면 좋을 것 같습니다.

LimdaeIl replied:
넵! 이거 다시 수정하도록 하겠습니다!

HanaHww2 reviewed (approved):
수고하셨습니다. (●'◡'●)

LimdaeIl commented on user/src/main/java/com/faster/user/app/user/infrastructure/persistence/jpa/dto/QUserQuerydslResponseDto.java:
```diff
@@ -0,0 +1,51 @@
+@Getter
+@Setter
+@NoArgsConstructor
+public class QUserQuerydslResponseDto {
```
감사합니다!! 와 이거 하나 더 배워갑니다..! 수정하도록 하겠습니다 !

LimdaeIl commented on user/src/main/java/com/faster/user/app/user/presentation/internal/InternalUserController.java:
```diff
@@ -0,0 +1,69 @@
+
+
+  @GetMapping
+  public ResponseEntity<ApiResponse<Page<QUserQuerydslResponseDto>>> getAllUsers(
```
가입된 모든 사용자인 경우 [회원 고유 식별 번호]를 통해서 단 건 [회원 상세 정보]를 조회하면 괜찮다고 생각했습니다..!
/api/ 로 수정할까요?

LimdaeIl replied:
@authcheck 도 반영하도록 하겠습니다!

HanaHww2 replied:
아하, 저는 search 검색이 마스터 사용자에게 필요한 기능이라고 생각을 했었거든요. 예를 들면, 어드민 페이지용이요! 내부 요청에서 따로 전체 사용자 조회할 일은 없을 거 같아서, api 로 가주셔도 좋을 거 같습니다.

je-pa commented on user/src/main/java/com/faster/user/app/user/application/usecase/UserServiceImpl.java:
```diff
@@ -31,4 +37,11 @@ public UpdateUserRoleResponseDto updateUserRoleByUserId(Long userId, UpdateUserR
+  @Override
+  public Page<QUserQuerydslResponseDto> getAllUsers(String username, String name, String slackId, Integer page, Integer size) {
+    Pageable pageable = PageRequest.of(page, size);
+    return userJpaRepository.searchUsers(username, name, slackId, pageable);
```
service 계층에서는 userJpaRepository 같은 infrastructure 계층을 직접적으로 사용하지 않기로 했어요!
관련해서 dip나 핵사고날 아키텍처 관련 공부 해보시면 좋을 것 같아요!

LimdaeIl replied:
오 감사합니다! 이것도 반영해서 다시 올리도록 하겠습니다!

je-pa reviewed (approved):
고생하셨습니다!

Commit:
[#75][user] 공통 응답으로 구현하기

- page -> pageResponse
- @AuthCheck 적용하기

Commit:
[#75][user] refactor: DIP 지키며 리포지토리 설계하기

커스텀 리포지토리랑 JPA 리포지토리 둘 다 사용하고 싶은 나는...

어댑터 클래스를 사용할 수 밖에 없었다!!!!!! JPA 기본 기능도 사용하고 싶다!!!!!

Commit:
[#75][user] refactor: DIP 수정에 따라 서비스에서 리포지토리 주입 수정하기

어댑터로 쿼리 디에셀이랑 JPA 둘 다 사용할 수 있도록 구현했다!
